rust/xaynet-client/src/utils/concurrent_futures.rs

```diff
@@ -0,0 +1,131 @@
+#![allow(dead_code)]
+
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{
+    stream::{FuturesUnordered, Stream},
+    Future,
+};
+use tokio::task::{JoinError, JoinHandle};
+
+/// `ConcurrentFutures` can keep a capped number of futures running concurrently, and yield their
+/// result as they finish. When the max number of concurrent futures is reached, new tasks are
+/// queued until some in-flight futures finish.
+pub struct ConcurrentFutures<T>
+where
+    T: Future + Send + 'static,
+    T::Output: Send + 'static,
+{
+    /// in-flight futures
+    running: FuturesUnordered<JoinHandle<T::Output>>,
+    /// buffered tasks
+    pending: VecDeque<T>,
+    /// max number of concurrent futures
+    max_in_flight: usize,
+}
+
+impl<T> ConcurrentFutures<T>
+where
+    T: Future + Send + 'static,
+    T::Output: Send + 'static,
+{
+    pub fn new(max_in_flight: usize) -> Self {
+        Self {
+            running: FuturesUnordered::new(),
+            pending: VecDeque::new(),
+            max_in_flight,
+        }
+    }
+
+    pub fn push(&mut self, task: T) {
+        self.pending.push_back(task)
+    }
+}
+
+impl<T> Unpin for ConcurrentFutures<T>
+where
+    T: Future + Send + 'static,
+    T::Output: Send + 'static,
+{
+}
+
+impl<T> Stream for ConcurrentFutures<T>
+where
+    T: Future + Send + 'static,
+    T::Output: Send + 'static,
+{
+    type Item = Result<T::Output, JoinError>;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        while this.running.len() < this.max_in_flight {
+            if let Some(pending) = this.pending.pop_front() {
+                let handle = tokio::spawn(pending);
+                this.running.push(handle);
+            } else {
+                break;
+            }
+        }
+        Pin::new(&mut this.running).poll_next(cx)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use futures::stream::StreamExt;
+    use tokio::time::delay_for;
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test() {
```

Comment:
just out of interest: can the timings of pushing futures always be guaranteed to be precise down to the millisecond? the timings in this test are quite tight, if pushing new futures after another has resolved is delayed for only 1ms (for example for test environments with heavy load), then this test will likely yield a different order of values.

Reply:
The interval between the futures is ~10ms, not 1ms, so I think it's enough for tests.

Reply:
the way i understood the timeline is the following: in the beginning all 4 futures are queued and none is spawned, then:
- 0ms: futures 1 and 2 are spawned, futures 3 and 4 are queued
- 10ms: future 1 resolves, future 2 is still pending, future 3 is spawned, future 4 is still queued
- 22ms: future 3 is resolved, future 2 is still pending, future 4 is spawned, none is queued any more
- 23ms: future 4 is resolved, future 2 is still pending
- 25ms: future 2 is resolved

this leaves only a two milliseconds between resolving the last two futures. if anything is delayed in between, this might be quite tight.

Reply:
Oh yes you're right. I should probably rework this a bit indeed.

Reply:
fixed.